src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultTagHelperFactory.cs

```diff
@@ -0,0 +1,107 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Internal
+{
+    /// <summary>
+    /// Default implementation for <see cref="ITagHelperFactory"/>.
+    /// </summary>
+    public class DefaultTagHelperFactory : ITagHelperFactory
+    {
+        private readonly ITagHelperActivator _activator;
+        private readonly ConcurrentDictionary<Type, PropertyActivator<ViewContext>[]> _injectActions;
+        private readonly Func<Type, PropertyActivator<ViewContext>[]> _getPropertiesToActivate;
+
+        /// <summary>
+        /// Initializes a new <see cref="DefaultTagHelperFactory"/> instance.
+        /// </summary>
+        /// <param name="activator">
+        /// The <see cref="ITagHelperActivator"/> used to create tag helper instances.
+        /// </param>
+        public DefaultTagHelperFactory(ITagHelperActivator activator)
+        {
+            if (activator == null)
+            {
+                throw new ArgumentNullException(nameof(activator));
+            }
+
+            _activator = activator;
+            _injectActions = new ConcurrentDictionary<Type, PropertyActivator<ViewContext>[]>();
+            _getPropertiesToActivate = type =>
+                PropertyActivator<ViewContext>.GetPropertiesToActivate(
+                    type,
+                    typeof(ViewContextAttribute),
+                    CreateActivateInfo);
+        }
+
+        /// <inheritdoc />
+        public TTagHelper CreateTagHelper<TTagHelper>(ViewContext context)
+            where TTagHelper : ITagHelper
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var tagHelper = _activator.Create<TTagHelper>(context);
+
+            var propertiesToActivate = _injectActions.GetOrAdd(
+                tagHelper.GetType(),
+                _getPropertiesToActivate);
+
+            for (var i = 0; i < propertiesToActivate.Length; i++)
+            {
+                var activateInfo = propertiesToActivate[i];
+                activateInfo.Activate(tagHelper, context);
+            }
+
+            InitializeTagHelper(tagHelper, context);
+
+            return tagHelper;
+        }
+
+        private static void InitializeTagHelper<TTagHelper>(TTagHelper tagHelper, ViewContext context)
+            where TTagHelper : ITagHelper
+        {
+            // Run any tag helper initializers in the container
+            var serviceProvider = context.HttpContext.RequestServices;
+            var initializers = (IEnumerable<ITagHelperInitializer<TTagHelper>>) 
```

Comment:
It was brought in transitively, intellisense was :trollface: me when I ctrl+. it